### PR TITLE
feat/metric_building

### PR DESCRIPTION
**Pull request by KrisMTLA**
It has been a while but we are back! Needed to refresh myself a bit on the original proposed metrics. Elaborated and built out 2 of the original 3 I scoped out. 

**Metric:**  customer age
Rather than simply deriving days between first and last order, I thought it would add value to create `days_since_first_order`, `days_since_last_order` and `days_between_first_and_last`. As well an `is_active` bool that is determined by a certain threshold `days_since_last_order`

I believe these metrics provide more value than the original proposal for the marketing team. 
`days_since_first_order` can be used to determine out longest standing customers for certain marketing analysis or potential campaigns 
`days_since_last_order` can help determine the current state of a customer, if they are current or not
`days_between_first_and_last` can help determine how long a customer has been making orders with our company

**Metric:**  time to deliver
Again, rather than simply providing time to deliver, I thought it would add value to add:
`days_from_ordered_to_carrier` - help determine how quickly we are getting orders out to our carriers 
`days_from_ordered_to_delivered` - help track total time to delivery
`days_from_carrier_to_delivered` - help track performance of the carrier
`order_delivery_status` - track how well we are delivering vs our estimates

### CODE REVIEW

File: models/marts/dim_customers.sql [hunk 1]
```diff
@@ -1,5 +1,34 @@
 with customers as  (
-    select * from {{ ref('stg_customers' )}}
+
```
Comment (JFrackson) [inline models/marts/dim_customers.sql:2]: Small comment: blank lines here aren't totally necessary but they can help with readability. Main thing is just consistency which you seem to be consistent about here so that's good!

File: models/marts/dim_customers.sql [hunk 2]
```diff
@@ -1,5 +1,34 @@
...
+    select * from {{ ref('stg_orders')}}
+
+),
+
+customer_first_last_order_at as (
+
+    select
+        customer_id,
+        first_value(orders.ordered_at) over(partition by orders.customer_id order by orders.ordered_at asc) as first_order_at,
+        last_value(orders.ordered_at) over(partition by orders.customer_id order by orders.ordered_at asc) as last_order_at
```
Comment (JFrackson) [inline models/marts/dim_customers.sql]: With SQLFluff we can enforce line length limits. Here is an example where it might be beneficial to break up these columns into multi-line statements. No need to change but just for you to think about when writing SQL

File: models/marts/dim_customers.sql [hunk 3]
```diff
@@ -13,10 +42,23 @@ final as (
...
         customers.customer_state,
+        days_since_first_last.days_since_first_order,
+        days_since_first_last.days_since_last_order,
+        days_since_first_last.days_between_first_last_order,
+    --boolean
+        --customer considered active if they have made an order in the past 45 days
+        case 
+            when days_since_first_last.days_since_last_order <= 45 then 1
+            else 0
+        end as is_active,
```
Comment (JFrackson) [inline models/marts/dim_customers.sql]: small thing: when I see is_ I assume the field will be true or false. 0 or 1 is also easily interpreted as a boolean, but it's worth thinking about: 1. the tool that will be using this (e.g. if it's Looker, 0 and 1 may be easier) and 2. naming patterns (e.g. if it's a boolean data type I could test the name to see if it starts with is_ or has_, but that test is not possible here). 

TLDR: true or false is easier to interpret because if I see  is_active==False I know better what it means than just seeing is_active==0. Small thing but worth thinking about when implementing booleans across a project

File: models/marts/fct_order_items.sql [hunk 4]
```diff
@@ -33,17 +41,35 @@ final as (
     -- dimensions
         order_items.price,
         order_items.freight_value,
-        case when order_payments.payment_type is null then 'returned'
-            else orders.order_status end as order_status,
+        case 
+            when order_payments.payment_type is null then 'returned'
+            else orders.order_status 
+        end as order_status,
```
Comment (JFrackson) [inline models/marts/fct_order_items.sql]: this "returned" bit of logic is not immediately clear: what if it's yet to be paid and therefore null? A comment would be helpful

File: models/marts/fct_order_items.sql [hunk 5]
```diff
@@ -33,17 +41,35 @@ final as (
...
         coalesce(order_payments.payment_type, 'returned') as payment_type,
         coalesce(order_payments.payment_installments, 0) as payment_installments,
         coalesce(order_payments.payment_value, 0) as payment_amount,
         reviews.review_score,
+        timestamp_diff(orders.order_delivered_carrier_date, orders.ordered_at, day) as days_from_ordered_to_carrier,
+        timestamp_diff(orders.order_delivered_customer_date, orders.ordered_at, day) as days_from_ordered_to_delivered,
+        timestamp_diff(orders.order_delivered_customer_date, orders.order_delivered_carrier_date, day) as days_from_carrier_to_delivered,
+        case
+            when orders.order_status = 'delivered' 
+                and extract(day from orders.order_delivered_customer_date) = extract(day from orders.order_estimated_delivery_date) then 'on time'
```
Comment (JFrackson) [inline models/marts/fct_order_items.sql]: the extract(day from ...) is a bit verbose and repetitive; a shorter alternative would be date(...). https://docs.snowflake.com/en/sql-reference/functions/to_date.html

Small thing but would make it more readable

File: models/marts/fct_order_items.sql [hunk 6]
```diff
@@ -33,17 +41,35 @@ final as (
...
+                and extract(day from orders.order_delivered_customer_date) < extract(day from orders.order_estimated_delivery_date) then 'early'
+        end as order_delivery_status,
+
     -- booleans
-        case when order_payments.payment_installments > 1 then 1 
-            else 0 end as is_paid_with_installments,
-        case when reviews.review_score is not null then 1 
-            else 0 end as is_reviewed,
+        case 
+            when order_payments.payment_installments > 1 then 1
```
Comment (JFrackson) [inline models/marts/fct_order_items.sql]: same boolean comment here

**Review by JFrackson — APPROVED** [review]
looks great! Left some specific style and naming convention comments to be didactic

File: models/marts/dim_customers.sql (hunk 2 above)
Comment (KrisMTLA) [reply models/marts/dim_customers.sql]: Modified

File: models/marts/dim_customers.sql (hunk 1 above)
Comment (KrisMTLA) [reply models/marts/dim_customers.sql:2]: noted thanks

File: models/marts/dim_customers.sql (hunk 3 above)
Comment (KrisMTLA) [reply models/marts/dim_customers.sql]: makes sense! till be consistent with creating true vs false bool fields moving forward

File: models/marts/fct_order_items.sql (hunk 4 above)
Comment (KrisMTLA) [reply models/marts/fct_order_items.sql]: agreed, not sure what my thought process was here tbh. There are 0 records with a null `order_status`
removed logic all together

File: models/marts/fct_order_items.sql (hunk 5 above)
Comment (KrisMTLA) [reply models/marts/fct_order_items.sql]: makes sense, moved logic to date()

File: models/marts/fct_order_items.sql (hunk 6 above)
Comment (KrisMTLA) [reply models/marts/fct_order_items.sql]: noted, modified my bools
